### ISSUE TITLE
Honeybadger.Metrics.Server crashes after upgrade deploy via Edeliver.

### ISSUE THREAD

**Issue by saaji** [honeybadger-io/honeybadger-elixir#44]
My Phoenix application uses Honeybadger 0.5.0 from Hex.

Application `honeybadger` is present in the list of applications. 

```
def application do
    [mod: {API, []},
     applications: [
       :phoenix,
       :phoenix_html,
       :cowboy,
       :logger,
       :gettext,
       :guardian,
       :connection,
       :phoenix_ecto,
       :postgrex,
       :tzdata,
       :timex,
       :timex_ecto,
       :httpoison,
       :cors_plug,
       :exometer_core,
       :exometer,
       :elixometer,
       :ex_rated,
       :exjsx,
       :edown,
       :parse_trans,
       :runtime_tools,
       :honeybadger,
       :edeliver]]
  end
```

The dependency list is

```
defp deps do
  [{:phoenix, "~> 1.1.4"},
   {:phoenix_ecto, "~> 2.0"},
   {:postgrex, ">= 0.0.0"},
   {:phoenix_html, "~> 2.5"},
   {:phoenix_live_reload, "~> 1.0", only: :dev},
   {:gettext, "~> 0.9"},
   {:cowboy, "~> 1.0"},
   {:exrm, "1.0.4"},
   {:edeliver, ">= 1.2.8"},
   {:exjsx, "~> 3.2"},
   {:timex, "~> 1.0"},
   {:timex_ecto, "~> 0.7.0"},
   {:guardian, "~> 0.9.0"},
   {:httpoison, "~> 0.8.0"},
   {:decimal, "~> 1.1.0"},
   {:cors_plug, "~> 0.1.4"},
   {:edown, "~>0.7.0", override: true},
   {:lager, "~>3.0.2", override: true},
   {:setup, "1.7.0", override: true},
   {:parse_trans, "2.9.0", override: true},
   {:exometer_core, github: "PSPDFKit-labs/exometer_core", override: true},
   {:exometer, github: "PSPDFKit-labs/exometer", override: true},
   {:elixometer, github: "pinterest/elixometer"},
   {:honeybadger, "~> 0.5"},
   {:ex_rated, "~> 1.2"}]
end
```

[eDeliver](https://github.com/boldpoker/edeliver) is being used to deploy and live-upgrade production environment via OTP releases. It uses exrm under the hood to do its job.

Honeybadger works well in case of 'cold' deploy and/or restarting the app. However, doing live upgrade via `mix edeliver upgrade production` makes `Honeybadger.Metrics.Server` to start crashing regularly with the following trace: 

```
08:13:52.918 [error] GenServer Honeybadger.Metrics.Server terminating
** (ArgumentError) argument error
    (honeybadger) lib/honeybadger/client.ex:28: Honeybadger.Client.send_metric/3
    (honeybadger) lib/honeybadger/metrics/server.ex:49: Honeybadger.Metrics.Server.handle_info/2
    (stdlib) gen_server.erl:615: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:681: :gen_server.handle_msg/5
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

=CRASH REPORT==== 24-Jun-2016::08:13:52 ===
  crasher:
    initial call: Elixir.Honeybadger.Metrics.Server:init/1
    pid: <0.14934.32>
    registered_name: 'Elixir.Honeybadger.Metrics.Server'
    exception exit: {badarg,
                        [{'Elixir.Honeybadger.Client',send_metric,3,
                             [{file,"lib/honeybadger/client.ex"},{line,28}]},
                         {'Elixir.Honeybadger.Metrics.Server',handle_info,2,
                             [{file,"lib/honeybadger/metrics/server.ex"},
                              {line,49}]},
                         {gen_server,try_dispatch,4,
                             [{file,"gen_server.erl"},{line,615}]},
                         {gen_server,handle_msg,5,
                             [{file,"gen_server.erl"},{line,681}]},
                         {proc_lib,init_p_do_apply,3,
                             [{file,"proc_lib.erl"},{line,240}]}]}
      in function  gen_server:terminate/7 (gen_server.erl, line 826)
    ancestors: [<0.1537.0>,<0.1536.0>]
    messages: []
    links: [<0.1537.0>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 610
    stack_size: 27
    reductions: 922
  neighbours:

=SUPERVISOR REPORT==== 24-Jun-2016::08:13:52 ===
     Supervisor: {<0.1537.0>,'Elixir.Honeybadger.Metrics.Supervisor'}
     Context:    child_terminated
     Reason:     {badarg,
                     [{'Elixir.Honeybadger.Client',send_metric,3,
                          [{file,"lib/honeybadger/client.ex"},{line,28}]},
                      {'Elixir.Honeybadger.Metrics.Server',handle_info,2,
                          [{file,"lib/honeybadger/metrics/server.ex"},
                           {line,49}]},
                      {gen_server,try_dispatch,4,
                          [{file,"gen_server.erl"},{line,615}]},
                      {gen_server,handle_msg,5,
                          [{file,"gen_server.erl"},{line,681}]},
                      {proc_lib,init_p_do_apply,3,
                          [{file,"proc_lib.erl"},{line,240}]}]}
     Offender:   [{pid,<0.14934.32>},
                  {id,'Elixir.Honeybadger.Metrics.Server'},
                  {mfargs,{'Elixir.Honeybadger.Metrics.Server',start_link,[]}},
                  {restart_type,permanent},
                  {shutdown,5000},
                  {child_type,worker}]
```

The only way to fix this is to restart the app completely either by hand on each node or by `mix deliver stop production` / `mix edeliver start production`.

Please note that my Honeybadger account does not have performance metrics enabled at all due to its subscription plan limitations.

**Comment by nurugger07** [comment]
Thanks @saaji for submitting this detailed issue. I'm going to look at it this weekend and see it I can get it fixed.

**Comment by parubets** [comment]
same problem for me, not only for Phoenix, but for all elixir apps.

`10:58:32.117 [error] Task #PID<0.27544.5> started from RatesApp.Kraken.Fetcher terminating
** (ArgumentError) argument error
    (honeybadger) lib/honeybadger/client.ex:36: Honeybadger.Client.send_notice/3
    (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: #Function<0.38809926/0 in RatesApp.Kraken.Fetcher.handle_info/2>
    Args: []

=CRASH REPORT==== 13-Jul-2016::10:58:32 ===
  crasher:
    initial call: Elixir.RatesApp.Kraken.Fetcher:-handle_info/2-fun-0-/0
    pid: <0.27544.5>
    registered_name: []
    exception exit: {badarg,
                        [{'Elixir.Honeybadger.Client',send_notice,3,
                             [{file,"lib/honeybadger/client.ex"},{line,36}]},
                         {'Elixir.Task.Supervised',do_apply,2,
                             [{file,"lib/task/supervised.ex"},{line,89}]},
                         {proc_lib,init_p_do_apply,3,
                             [{file,"proc_lib.erl"},{line,240}]}]}
      in function  'Elixir.Task.Supervised':exit/4 (lib/task/supervised.ex, line 120)
    ancestors: ['Elixir.RatesApp.Kraken.Fetcher',
                  'Elixir.RatesApp.Supervisor',<0.843.0>]
    messages: []
    links: []
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 12037
  neighbours:

10:58:32.119 [error] Task #PID<0.27545.5> started from :error_logger terminating
** (ArgumentError) argument error
    (honeybadger) lib/honeybadger/client.ex:36: Honeybadger.Client.send_notice/3
    (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: #Function<0.10127598/0 in Honeybadger.Logger.handle_event/2>
    Args: []

=CRASH REPORT==== 13-Jul-2016::10:58:32 ===
  crasher:
    initial call: Elixir.Honeybadger.Logger:-handle_event/2-fun-0-/0
    pid: <0.27545.5>
    registered_name: []
    exception exit: {badarg,
                        [{'Elixir.Honeybadger.Client',send_notice,3,
                             [{file,"lib/honeybadger/client.ex"},{line,36}]},
                         {'Elixir.Task.Supervised',do_apply,2,
                             [{file,"lib/task/supervised.ex"},{line,89}]},
                         {proc_lib,init_p_do_apply,3,
                             [{file,"proc_lib.erl"},{line,240}]}]}
      in function  'Elixir.Task.Supervised':exit/4 (lib/task/supervised.ex, line 120)
    ancestors: [error_logger,<0.2.0>]
    messages: []
    links: []
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 2586
    stack_size: 27
    reductions: 13500
  neighbours:

10:58:32.122 [error] Task #PID<0.27546.5> started from :error_logger terminating
** (ArgumentError) argument error
    (honeybadger) lib/honeybadger/client.ex:36: Honeybadger.Client.send_notice/3
    (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: #Function<0.10127598/0 in Honeybadger.Logger.handle_event/2>
    Args: []
`

**Comment by joshuap** [comment]
This should be fixed once we remove the metrics server (see http://blog.honeybadger.io/sunsetting-performance-metrics/)

**Comment by parubets** [comment]
have you removed already in 0.6.1?

**Comment by joshuap** [comment]
Not yet. We'll update this issue when this fix has been released.

**Comment by nurugger07** [comment]
This isn't available in the 0.6.2 release but we'll be doing a major release later this week that will be removing metrics reporting. Since it is a major change, I wanted to do a minor release first

**Comment by mwoods79** [comment]
Metrics are removed in https://github.com/honeybadger-io/honeybadger-elixir/pull/80. @saaji will you install from github and verify the problem is fixed?

**Comment by joshuap** [comment]
I'm going to close this now that we've released 0.6.3 (which removed metrics). Feel free to re-open/submit a new issue if you continue to have problems.
